api/logs/v1/rules_labels_enforcer.go

```diff
@@ -0,0 +1,179 @@
+// rules are filtered by labels applied by the user.
+func WithEnforceRuleLabelFilters(logger log.Logger, labelKeys []string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			level.Debug(logger).Log("msg", "filtering by labels", "keys", fmt.Sprintf("%s", labelKeys))
+
+			if len(labelKeys) == 0 {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			f := &responseLabelFilter{ResponseWriter: w}
+
+			next.ServeHTTP(f, r)
+
+			matchers := extractMatchers(r, labelKeys)
+
+			contentType := w.Header().Get("Content-Type")
+
+			previous := fmt.Sprint(len(f.body))
+
+			matcherStr := fmt.Sprintf("%v", matchers)
+			level.Debug(logger).Log("msg", "filtering by matchers", "matchers", matcherStr)
+
+			b, err := filterRules(f.body, contentType, matchers)
+			if err != nil {
+				level.Error(logger).Log("msg", err)
+				f.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+
+			current := fmt.Sprint(len(b))
+
+			level.Debug(logger).Log("msg", "filtered rules by", "previous len", previous, "current len", current)
+
+			f.Header().Set("Content-Length", fmt.Sprint(len(b)))
+
+			_, err = f.Write(b)
+			if err != nil {
+				f.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+		})
+	}
+}
+
+func extractMatchers(r *http.Request, l []string) map[string]string {
+	queryParams := r.URL.Query()
+	matchers := map[string]string{}
+	for _, name := range l {
+		value := queryParams.Get(name)
+		if value != "" {
+			matchers[name] = value
+		}
+	}
+
+	return matchers
+}
+
+func filterRules(body []byte, contentType string, matchers map[string]string) ([]byte, error) {
+	switch contentType {
+	case "application/json":
+		var res prometheusRulesResponse
+		err := json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		res.Data = filterPrometheusRules(res.Data, matchers)
+
+		return json.Marshal(res)
+
+	case "application/yaml":
+		var res lokiRulesResponse
+		err := yaml.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		return yaml.Marshal(filterLokiRules(res, matchers))
+
+	default:
+		return nil, errUnknownRulesContentType
+	}
+}
+
+func filterPrometheusRules(rd rulesData, matchers map[string]string) rulesData {
+	if len(matchers) == 0 {
+		return rd
+	}
+
+	var filtered []rulefmt.RuleGroup
+
+	for _, rg := range rd.RuleGroups {
+		var filteredRules []rulefmt.RuleNode
+
+	rules:
+		for _, rule := range rg.Rules {
+			for key, value := range matchers {
+				val, ok := rule.Labels[key]
+				if !ok || val != value {
+					continue rules
+				}
+			}
+
+			filteredRules = append(filteredRules, rule)
+		}
+
+		if len(filteredRules) > 0 {
+			filtered = append(filtered, rulefmt.RuleGroup{
+				Name:     rg.Name,
+				Interval: rg.Interval,
+				Limit:    rg.Limit,
+				Rules:    filteredRules,
+			})
+		}
+	}
+
+	return rulesData{RuleGroups: filtered}
+}
```

Comment:
Here's an alternative implementation that's easier to read:
```suggestion
func filterPrometheusRules(rd rulesData, matchers map[string]string) rulesData {
	if len(matchers) == 0 {
		return rd
	}

	var filtered []rulefmt.RuleGroup

	for _, rg := range rd.RuleGroups {
		var filteredRules []rulefmt.RuleNode

		for _, rule := range rg.Rules {
			if ruleLabelsMatcheMatchers(rule, matchers) {
				filteredRules = append(filteredRules, rule)
			}
		}

		if len(filteredRules) > 0 {
			filtered = append(filtered, rulefmt.RuleGroup{
				Name:     rg.Name,
				Interval: rg.Interval,
				Limit:    rg.Limit,
				Rules:    filteredRules,
			})
		}
	}

	return rulesData{RuleGroups: filtered}
}

func ruleLabelsMatcheMatchers(rule rulefmt.RuleNode, matchers map[string]string) bool {
	for key, value := range matchers {
		val, ok := rule.Labels[key]
		if !ok || val != value {
			return false
		}
	}
	return true
}
```